devito/finite_differences/tools.py

```diff
@@ -47,6 +46,7 @@ def adjoint(self, matvec):
 def check_input(func):
     @wraps(func)
     def wrapper(expr, *args, **kwargs):
+        from devito.finite_differences import Differentiable
         if expr.is_Number:
             return S.Zero
         elif not isinstance(expr, Differentiable):
```

Comment:
@Leitevmd @mloubout 

To avoid the circular import we should stop importing Differentiable into tools.

To avoid using Differentiable in tools, we could turn this `if` into a `try-except`, thus making it more pythonic, and potentially much much faster, since this function is likely to be called loads of times, and `isinstance` is relatively slow, enough to be a PITA if called too often

Mathias, how about we just do:

```
try:
    return func(expr, *args, **kwargs)
except ValueError:  # Not sure if a better exception should be picked; run some tests and find out?
    raise ValueError("`%s` must be of type Differentiable, not `%s`" (expr, type(expr))
```

?

Reply:
Tested! When no `Differentiable` is passed, it later raises` AttributeError`. So I'm gonna use it instead of `ValueError` so that we can see this step, right?